contracts/DsaWrapper.sol

```diff
@@ -1,62 +1,244 @@
-        bytes4 basicDeposit = bytes4(keccak256("deposit(address,uint256,uint256,uint256)"));
+        bytes4 basicWithdraw = bytes4(
+            keccak256("withdraw(address,uint256,address,uint256,uint256)")
+        );
 
         address tokenAddress = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
 
-        data[0] = abi.encodeWithSelector(basicDeposit, tokenAddress, msg.value, 0, 0);
+        data[0] = abi.encodeWithSelector(
+            basicWithdraw,
+            tokenAddress,
+            _amt,
+            msg.sender,
+            0,
+            0
+        );
 
-        console.log(abc.isAuth(msg.sender)); // false
-        console.log(abc.isAuth(address(this))); // false
+        IDSA(dsaAddress).cast(targets, data, address(this));
+    }
+
+    /**
+     * @dev deposits ERC20 token to DSA account given by _id
+     * @param _id id of DSA account
+     * @param _amt amount of tokens to be deposited
+     * @param _tokenAddress address of token to be deposited
+     */
+    function depositErc20(
+        uint256 _id,
+        uint256 _amt,
+        address _tokenAddress
+    ) external payable onlyAuthority(_id, msg.sender) {
+        IERC20 token = IERC20(_tokenAddress);
 
-        IDSA(dsaAddress).cast{value: msg.value}(targets, data, address(0));
+        // get dsa-address
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+
+        token.safeTransferFrom(msg.sender, address(this), _amt);
+        token.safeApprove(dsaAddress, _amt);
+
+        string[] memory targets = new string[](1);
+        targets[0] = "BASIC-A";
+
+        bytes[] memory data = new bytes[](1);
+        bytes4 basicDeposit = bytes4(
+            keccak256("deposit(address,uint256,uint256,uint256)")
+        );
+
+        data[0] = abi.encodeWithSelector(
+            basicDeposit,
+            _tokenAddress,
+            _amt,
+            0,
+            0
+        );
+        IDSA(dsaAddress).cast(targets, data, address(this));
+    }
+
+    /**
+     * @dev withdraws ERC20 token from DSA account given by _id
+     * @param _id id of DSA account
+     * @param _amt amount of tokens to withdraw
+     * @param _tokenAddress contract address of ERC20 token
+     */
+    function withdrawErc20(
+        uint256 _id,
+        uint256 _amt,
+        address _tokenAddress
+    ) external onlyAuthority(_id, msg.sender) {
+        IERC20 token = IERC20(_tokenAddress);
+        
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+
+        require(token.balanceOf(dsaAddress) >= _amt, "INSUFFICIENT TOKEN BALANCE");
+
+        string[] memory targets = new string[](1);
+        targets[0] = "BASIC-A";
+
+        bytes[] memory data = new bytes[](1);
+        bytes4 basicWithdraw = bytes4(
+            keccak256("withdraw(address,uint256,address,uint256,uint256)")
+        );
+
+        data[0] = abi.encodeWithSelector(
+            basicWithdraw,
+            _tokenAddress,
+            _amt,
+            msg.sender,
```

Comment:
Shouldn't this be `dsaAddress`?
Why withdraw at `msg.sender`?

Reply:
https://docs.instadapp.io/connectors/mainnet/basic#withdraw
The third field denotes "the address to receive the token upon withdrawal", so it should be `msg.sender`, right?

Reply:
Ok, my bad.